### ISSUE TITLE
Major refinements regressions in 3.9.x

### ISSUE THREAD

**Issue by m-rutter** [colinhacks/zod#688]
Refinements have regressed in 3.9.x
```ts
import { z } from 'zod';

z
  .string()
  .refine(() => false)
  .transform(() => {
    throw new Error("this shouldn't happen") // in 3.9.x this throws, in 3.8.x it doesn't reach this branch
  })
  .safeParse("") 
  ```

3.9.x - throws:
https://codesandbox.io/s/angry-hawking-qqx2q?file=/src/index.ts
 
3.8.x - doesn't throw:
https://codesandbox.io/s/elastic-hooks-eglbu?file=/src/index.ts

**Comment by Djaler** [comment]
It's happen here https://github.com/colinhacks/zod/commit/61c47f46259c46a8cd7785e80c813cbed90d5829

**Comment by Djaler** [comment]
@colinhacks can you explain a reasons?

**Comment by FlorianWendelborn** [comment]
@Djaler I can’t speak for @colinhacks, but from my perspective this just looks like a minor mistake. Basically, 3.9 is trying to gather as many validation errors as possible, and uses some heuristics to guess when to continue parsing in order to gather more errors. (this is very useful for e.g. form validations presented to end-users)

So, the point here would be that after the `.transform` there might be another `.refine` etc. that has more errors.

Personally, I think due to the risky nature of `.transform` (zod afaik has no idea what happens inside) the heuristic algorithm shouldn’t continue with a `.transform()` after encountering a falsy `.refine()`.

This is also sort-of documented/implied to be possible in the readme:

> ⚠️ Transform functions must not throw. **Make sure to use refinements before the transform to make sure the input can be parsed by the transform.**

**Comment by m-rutter** [comment]
I really like the library, but I have to say the 3.9.x branch has been a little rough. 

Combined with this: https://github.com/colinhacks/zod/issues/679 and looking through the codebase I've seen a lot of commented out tests that should have caught these changes. 

How does something like this slip into a release?
```
        // don't abort early on refinement issues
        // invalid = true;
```

**Comment by Djaler** [comment]
Yeah, I think we need to increase test coverage. If you interested in this, I can configure Stryker (mutation testing framework)

**Comment by colinhacks** [comment]
This was a rough release, made worse because I've been out on a personal trip for the last five days. I'll fix this shortly. 

@m-rutter While technically that's the source of the bug, that comment is very intentional. Here `invalid` means "fatal error has occurred, abort early". As @FlorianWendelborn said, the nature of the 9.3 change is that failed refinements don't abort early which maximizes the number of errors that can be surfaced to the user on the initial check. 

The oversight here is that Zod needs to check if any refinements have failed higher in the chain before executing and transforms. Everyone is right that there should have been a test for this.

**Comment by colinhacks** [comment]
This should be patched in 3.9.7

**Comment by Djaler** [comment]
@colinhacks so what about mutation testing?
